app/models/course_template.rb

```diff
@@ -47,13 +47,25 @@ def self.new_dummy(course, options = {})
   end
 
   def valid_git_repo?
-    return true unless source_url_changed? || git_branch_changed? # don't attempt repo cloning if source url or branch wasn't even changed
-    Dir.mktmpdir do |dir|
-      sh!('git', 'clone', '-q', '-b', git_branch, source_url, dir)
-      File.exist?("#{dir}/.git")
+    timeout = 10
+    safe_source_url = Shellwords.escape(source_url)
+    safe_git_branch = Shellwords.escape(git_branch)
+    cmd = "timeout #{timeout+1} git ls-remote --exit-code #{safe_source_url} #{safe_git_branch}"
+    output = ''
+    begin
+      Timeout::timeout(timeout) do
+        output = `#{cmd} 2>&1`
+      end
+    rescue Timeout::Error
+      errors.add(:base, "Cannot clone repository, timeout expired, git ls-remote output: #{output}")
+      return false
```

Comment:
As discussed with @mpartel last week, we probably should move this to the SystemCommands class and keep all escapes there.